### PR TITLE
Time value equals hash code

### PR DESCRIPTION
**Pull request by garydgregory**
@michael-o Please review.

### CODE REVIEW

**Comment by michael-o** [comment]
Will look into it tomorrow. Thanks!

File: httpcore5/src/main/java/org/apache/hc/core5/util/BigIntegerTimeUnit.java [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+            return d;
+        }
+
+        @Override
+        public BigInteger toMicros(final BigInteger d) {
+            return d.divide(C1.divide(C0));
+        }
+
+        @Override
+        public BigInteger toMillis(final BigInteger d) {
```
Comment (michael-o) [inline httpcore5/src/main/java/org/apache/hc/core5/util/BigIntegerTimeUnit.java]: Would you mind to stick to the given convention: `<prefix><unit>`?

**Comment by michael-o** [comment]
@ok2c It actually does not, you can all solve it in the given class. Gary has chosen this approach. The benefit is that this enum still remains package private (an implementation detail).
I also agree, reformatting and logical changes don't go well together.

**Comment by michael-o** [comment]
@garydgregory The tests now pass, But I think this needs to be folded into the `TimeValue` class itself.

**Comment by garydgregory** [comment]
Hi @michael-o 

I mimicked `BigIntegerTimeUnit` after `TimeUnit`, hence the method names. Are you saying you'd like to see `toMillis()` and friends renamed to `toMilliseconds()`? 

Also, what do you think about a better class name like `LosslessTimeUnit`, or something else that conveys the lack of loss of precision like `UnboundTimeUnit`?

**Comment by michael-o** [comment]
@garydgregory I do not really understand why a lossless time value class is necessary. Why not simply scale within the equals method with a private method? By scale I mean to multiply with the smallest prefix. E.g., `"10_000 ms".equals("10 s")` => `10 x ms_in_s` => compare. Similar to `compareTo()`

**Comment by garydgregory** [comment]
Tangent for a separate PR: `TimeValue` could implement `Comparable`.

**Comment by garydgregory** [comment]
> @garydgregory I do not really understand why a lossless time value class is necessary. Why not simply scale within the equals method with a private method? By scale I mean to multiply with the smallest prefix. E.g., `"10_000 ms".equals("10 s")` => `10 x ms_in_s` => compare. Similar to `compareTo()`

Hi @michael-o ,
That's just a different way to do it. Personally, I would not bother looking for the 'complication' ;-) of 'finding the smallest scale', I would just _normalize_ on the smallest scale and be done. In my mind, I wanted to cleanly separate the fact that the stock TimeValue class has this built-in boundary that gets in the way. The new class cleanly separates out this concern. That's IMO, YMMV :-)

**Comment by michael-o** [comment]
@

> 
> 
> > @garydgregory I do not really understand why a lossless time value class is necessary. Why not simply scale within the equals method with a private method? By scale I mean to multiply with the smallest prefix. E.g., `"10_000 ms".equals("10 s")` => `10 x ms_in_s` => compare. Similar to `compareTo()`
> 
> Hi @michael-o ,
> That's just a different way to do it. Personally, I would not bother looking for the 'complication' ;-) of 'finding the smallest scale', I would just _normalize_ on the smallest scale and be done. In my mind, I wanted to cleanly separate the fact that the stock TimeValue class has this built-in boundary that gets in the way. The new class cleanly separates out this concern. That's IMO, YMMV :-)

I am fine with scaling down to nanoseconds. This will do easily, `long` should be large enough. How you scale is an implementation detail. I do not see a huge need for a separate class for simple math.

**Comment by ok2c** [comment]
@garydgregory @michael-o I see no justification for the use of BigInteger math here.

**Comment by michael-o** [comment]
@ok2c I agree.

**Comment by ok2c** [comment]
Superseded by #172

**Pull request closed without merging**